Add 'Learn' guides (#48)
OpenZeppelin/openzeppelin-docs: components/learn/modules/ROOT/pages/write-contracts.adoc

= Write Solidity Contracts

Welcome to the exciting world of smart contract development! This guide will let you get started writing Solidity contracts by going over the following:

 * <<setting-up-a-solidity-project, Setting up a Solidity Project>>
 * <<compiling-solidity-source-code, Compiling Solidity Source Code>>
 * <<multi-file-projects, Multi-File Projects>>
 * <<using-openzeppelin-contracts, Using the OpenZeppelin Contracts library>>

=== About Solidity

We won't cover language concepts such as syntax or keywords in this guide. For that, you'll want to check out the following curated content, which feature great learning resources for both newcomers and experienced developers:

>  * For a general overview of how Ethereum and smart contracts work, the official website hosts a https://ethereum.org/learn/[Learn about Ethereum] section with lots of beginner-friendly content

Reviewer: ```suggestion
 * For a general overview of how Ethereum and smart contracts work, the official website hosts a https://ethereum.org/learn/[Learn about Ethereum] section with lots of beginner-friendly content, including an https://studio.ethereum.org/[online IDE] to write your own smart contracts projects.
```

Author: I didn't include the IDE on purpose: not only do I find it confusing, but it also uses the Contracts library precisely the way the teach our users not to, by copy-pasting the code.

 * If you're new to the language, the https://solidity.readthedocs.io/en/latest/introduction-to-smart-contracts.html[official Solidity documentation] is a good resource to have handy. Take a look at their https://solidity.readthedocs.io/en/latest/security-considerations.html[security recommendations], which nicely go over the differences between blockchains and traditional software platforms
 * These https://consensys.github.io/smart-contract-best-practices/[best practices] are quite extensive, and include both https://consensys.github.io/smart-contract-best-practices/recommendations/[proven patterns] to learn from and https://consensys.github.io/smart-contract-best-practices/known_attacks/[known pitfalls] to avoid
 * The https://solidity-05.ethernaut.openzeppelin.com/[Ethernaut] web-based game will have you look for subtle vulnerabilities in smart contracts as you advance through levels of increasing difficulty

With that out of the way, let's get started!

[[setting-up-a-solidity-project]]
== Setting up a Solidity Project

> The first step in any project is to create a `contracts` directory, which is where you will store your Solidity source files (`.sol`). This is equivalent to the `src` directory you may be familiar with from other languages.

Reviewer: ```suggestion
The first step in any project is to create a `contracts` directory, which is where we will store our Solidity source files (`.sol`). This is equivalent to the `src` directory you may be familiar with from other languages.
```

We can now write our first simple smart contract:

> ```solidity

Reviewer: Same as in the previous guide: I'd be explicit as to where this should be placed, either in the previous paragraph or as in a comment in the first line of the code block.

pragma solidity ^0.5.0;

contract SetAndGet {
    uint256 private value;

    // Emitted when the stored value changes
    event ValueChanged(uint256 newValue);

    // Stores a new value in the contract
    function set(uint256 newValue) public {
      value = newValue;

      emit ValueChanged(newValue);
    }

    // Reads the last stored value
    function get() public view returns (uint256) {
      return value;
    }
}
```

> We will store it in `contracts/StoreAndGet.sol`. Each `.sol` file should have the code for a single contract, and be named after it.

Reviewer: Hah, I should have scrolled further down before commenting. Still, I'd include it before (not after) the code block.

> We will store it in `contracts/StoreAndGet.sol`. Each `.sol` file should have the code for a single contract, and be named after it.

Reviewer: > Each `.sol` file should have the code for a single contract, and be named after it.

Maybe clarify that this is just by convention, though not enforced?

Author: I'd skip this detail and just teach people how to do things the _right_ way.

[[compiling-solidity-source-code]]
== Compiling Solidity Source Code

The Ethereum Virtual Machine (EVM) cannot execute Solidity code directly: we first need to compile it into EVM bytecode.

To do so, we'll use the https://docs.openzeppelin.com/sdk/2.6/[*OpenZeppelin command-line interface (CLI)*]:

```bash
$ npm install --save-dev @openzeppelin/cli
```

> The CLI is quite powerful, and it is the entry point to many exciting features (such as upgradeable smart contracts and interactive transactions!). We will come back to it often.

Reviewer: ```suggestion
The CLI is quite powerful, and it is the entry point to many exciting features (such as upgradeable smart contracts and interactive transactions). We will come back to it often.
```

This is just me being boring. Feel free to disregard.

Author: It literally says _exciting_ in that same sentence :joy:

For now however, let's focus on compiling. This can be achieved by running a single command:

```bash
$ npx oz compile
```

NOTE: If you're unfamiliar with the `npx` command, check out our https://example.com[project setup guide].

The `compile` command will automatically look for all contracts in the `contracts` directory, and compile them using the Solidity compiler version that best suits your project.

> You will notice a `build/contracts` directory was created: it holds the compiled artifacts (bytecode and metadata), which are `.json` files. It's a good idea to add this directory to your `.gitignore`.

Reviewer: Should we add a one-line code snippet of `echo "build/contracts" >> .gitignore`?

Author: I thought about this, but felt this might side-track people a bit. I'd review this once we settle on contents for the 'dev environment' guide.

You can also configure your compilation, which includes choosing a compiler version and enabling optimizations, by passing arguments to the `compile` command:

```bash
$ npx oz compile --solc-version=0.5.12 --optimizer on
```

For detailed information on these options, refer to the https://docs.openzeppelin.com/sdk/2.6/api/cli#compile[`compile` API reference].

[[multi-file-projects]]

> == Multi-File Projects

Reviewer: I'd look for a simpler title. Maybe something like "Adding more contracts"?

> There are multiple reasons that cause most projects to have multiple Solidity source files: from modularization via https://solidity.readthedocs.io/en/latest/contracts.html#inheritance[inheritance] to complex multi-contract systems.

Reviewer: Let's tone this down. We can just state that _as our project grows, we will need to create more contracts, which may interact with each other_.

Author: 'Tone down' as in make it simpler?

Reviewer: Yes, _multi-file projects_ sounds intimidating, or even enterprisey. We are just talking about throwing in more files to our project.

Because the compiler will pick up all files in the `contracts` directory and subdirectories, you are free to organize your code as you see fit (such as having subdirectories for `access-control` or `token` contracts).

To import a contract from another file, refer to it by its relative path using an `import` statement:

```solidity
import './Math.sol'; // Import from the same directory

> import './access-control/Admin.sol'; // Import from a subdirectory

Reviewer: WDYT about actually creating another contract, even if ridiculously simple, to import it? Otherwise it may be weird to see these two new names just spring up. We could write sth that uses the SetAndGet for instance!

```

[[using-openzeppelin-contracts]]
== Using the OpenZeppelin Contracts Library

> Reusable modules and libraries are the cornerstone of great software. The https://openzeppelin.com/contracts/[*OpenZeppelin Contracts*] library contains lots of useful building blocks for smart contracts to build on. And you can rest easy: they've been the subject of multiple security audits, with their security and correctness battle-tested.

Reviewer: >Reusable modules and libraries are the cornerstone of great software

In your face, Vyper!

Nothing to change here, just wanted to add that.

> Reusable modules and libraries are the cornerstone of great software. The https://openzeppelin.com/contracts/[*OpenZeppelin Contracts*] library contains lots of useful building blocks for smart contracts to build on. And you can rest easy: they've been the subject of multiple security audits, with their security and correctness battle-tested.

Reviewer: ```suggestion
Reusable modules and libraries are the cornerstone of great software. The https://openzeppelin.com/contracts/[*OpenZeppelin Contracts*] library contains lots of useful building blocks for smart contracts to build on. And you can rest easy: they've been the subject of multiple audits, with their security and correctness battle-tested.
```

To use the library, you'll first need to download it from the latest published release:

```bash
$ npm install --save-dev @openzeppelin/contracts
```

NOTE: You should always use the library from these published releases: copy-pasting library source code into your project makes it very easy to introduce security vulnerabilities in your contracts.

> To use one of the OpenZeppelin Contracts, `import` its path with `@openzeppelin/contracts`.

Reviewer: > `import` its path with `@openzeppelin/contracts`

This bit is unclear. Let's try and rephrase it!

Author: Huh, looks like I deleted some extra words here in the edit. Thanks!

> Let's see how limiting access control to our `SetAndGet` contract looks like:

Reviewer: This part is difficult to understand without having prior knowledge of Ownable. I like the intent of adding the explanation in a comment on the import itself, but tbh I had initially glossed over it and completely missed it. I'd add here a little bit about OZ contracts, given it's the first time we're introducing it, and explain that it has out-of-the box contracts (eg erc20) as well as aspects that you can inject into your contracts, then explaining multiple inheritance as a mechanism for code reuse in Solidity. And then I'd go into presenting Ownable, and this example.

Author: I tried to improve this, but think I may have gone a bit too far :sweat_smile:

[[set-and-get-contract]]

> ```solidity

Reviewer: Again, let's include the path, so readers can easily understand that this is the same file

// giving it special permissions.
import '@openzeppelin/contracts/ownership/Ownable.sol';

contract SetAndGet is Ownable { // Inherit from the Ownable contract
    uint256 private value;

    event ValueChanged(uint256 newValue);

    // Stores a new value in the contract
    function set(uint256 newValue) public onlyOwner { // Restrict access to set
      value = newValue;

      emit ValueChanged(newValue);
    }

    // Reads the last stored value
    function get() public view returns (uint256) {
      return value;
    }
}
```

Despite the OpenZeppelin Contracts not being located in your `contracts` directory, the CLI is smart enough to know where to look for them while compiling.

=== More on OpenZeppelin Contracts

The https://docs.openzeppelin.com/contracts/2.x/#next-steps[OpenZeppelin Contracts documentation] is a great place to learn about developing secure smart contract systems. It features both guides and a detailed API reference: see for example the https://docs.openzeppelin.com/contracts/2.x/access-control[Access Control guide] to know more about the `Ownable` contract used in the code sample above.

> While the library is designed in such a way that it is difficult to misuse it, when in doubt you can always refer to the documentation for clarification.

Reviewer: ```suggestion
```

Show, don't tell ;-)

== Next Steps

Writing and compiling Solidity contracts are but the first steps in the journey to having your decentralized application running on the Ethereum network. Once you are confortable with this setup, you'll want to move on to more advanced tasks:

>  * xref:unit-testing.adoc[Perform automated contract testing]

Reviewer: ```suggestion
 * xref:unit-testing.adoc[Testing your contracts]
```
This is something I got from a book on writing. We're tending towards a noun-ification of language, which only adds boilerplate. Test is already a verb, which is turned into a noun (testing), and turned again into an action (perform testing). We should aim for simplicity, and put the focus on actions (verbs), not on things (nouns).

Author: Interesting! Can you share more about that?

Reviewer: See Slack ;-)

> = Write Solidity Contracts

Reviewer 2: I was a bit disappointed that the title is "write Solidity" yet the guide doesn't cover anything about actually writing Solidity. The next paragraph even says it will help me get started writing Solidity contracts but it's kind of far from that.

We should make sure that the title matches the content so that the reader's expectations are met.

Author: Yes, I should've mentioned that all titles are work in progress. I think these guides will be easier to name once we have a more complete picture, and how they relate to each other

Reviewer: I'd just go with "Writing smart contracts"

> Welcome to the exciting world of smart contract development! This guide will help get you started writing Solidity contracts, importing libraries, and compiling these into deployable Ethereum bytecode. Onwards!

Reviewer 2: I think this summary at the beginning of the page is great. We should definitely do it everywhere. What do you think about using an unordered list, with one item per line though? It will be easier to parse.

== Learn about Solidity

> We won't cover language concepts, syntax, etc. in this guide. For that, you'll want to check out the following curated content, which features great learning resources for new developers and veterans alike:

Reviewer 2: I don't think "etc." reads well. Can we adopt the rules suggested [here](https://insidegovuk.blog.gov.uk/2016/07/20/changes-to-the-style-guide-no-more-eg-and-ie-etc/)? I also agree with the article on discouraging "e.g.", though I feel that sometimes (e.g., in parentheses) it can be ok.

Author: I really like this, thanks!

>  * For a general overview of how Ethereum and smart contracts work, you can't go wrong with Matt Condon's https://medium.com/@mattcondon/getting-up-to-speed-on-ethereum-63ed28821bbe[Getting Up to Speed on Ethereum]

Reviewer: I'd swap this out in favour of the official https://ethereum.org/build/ page

 * A set of https://consensys.github.io/smart-contract-best-practices/[best practices] was compiled by Consensys Dilligence: these are quite extensive, and include both https://consensys.github.io/smart-contract-best-practices/recommendations/[proven patterns] to learn from and https://consensys.github.io/smart-contract-best-practices/known_attacks/[known pitfalls] to avoid
 * A rite of passage in the learning of Solidity is playing our very own https://solidity-05.ethernaut.openzeppelin.com/[Ethernaut] game: it'll test your understanding of both language and Ethereum concepts by having you look for subtle vulnerabilities in smart contracts as you advance through the increasingly difficult levels

== Setting up a Solidity Project

Simple works best: all you need to do is create directory called `contracts` and store your Solidity source files (`.sol`) there. That's it!

All tools will automatically pick every contract in this directory, letting you organize your code in subdirectories (e.g. `admin`, `token`, etc.). To import a file from a different directory, simply refer to it by its relative path:

```solidity
import './admin/TransferrableAdmin.sol';
```

> == Using the OpenZeppelin Contracts Library

Reviewer: I'd reshuffle the order in this guide:

1. Create a contracts folder 
2. Create a solidity file (with content!)
3. Compile it
4. Create another file and import it
5. Import contracts

Reusable modules and libraries are the cornerstone behind great software. The https://openzeppelin.com/contracts/[*OpenZeppelin Contracts*] library is the result of years of community-driven effort towards a set of useful building blocks for smart contracts to build on. They've been the subject of multiple security audits, and their security and correctness battle-tested. And using them is as easy as any other contract!

You'll first need to download the contracts from the latest published release:

```bash
$ npm install --save-dev @openzeppelin/contracts
```

Once installed, the library is ready to be used! Simply `import` the contracts as you would any other source in your system, prefixing the path with `@openzeppelin/contracts`:

```solidity
pragma solidity ^0.5.0;

import '@openzeppelin/contracts/ERC20/ERC20.sol';

contract MyToken is ERC20 {
  constructor(uint256 initialSupply) public {
    _mint(msg.sender, initialSupply);
  }
}
```

> The OpenZeppelin Contracts feature extensive https://docs.openzeppelin.com/contracts/2.x/#next-steps[documentation], complete with a detailed API reference: see for example the entry for the https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#ERC20-_mint-address-uint256-[`_mint` function] used in the code sample above.

Reviewer: Less shilling!

While the library is designed in such a way that it is difficult to misuse it, when in doubt, you can always refer to the documentation for clarification.

== Compiling your Contracts

> Your contracts are set up, libraries installed, but how do you actually transform these Solidity source files into executable Ethereum bytecode?

Reviewer 2: It's not obvious that "transform into executable bytecode" is what "compile" means. I would add a sentence saying something like "before we deploy we have to compile the source code into EVM bytecode".
